src/go/k8s/pkg/resources/certmanager/pki.go

```diff
@@ -0,0 +1,155 @@
+}
+
+// NodeCert returns the namespaced name for Redpanda's node certificate
+func (r *PkiReconciler) NodeCert() types.NamespacedName {
+	return types.NamespacedName{Name: r.pandaCluster.Name + "-" + RedpandaNodeCert, Namespace: r.pandaCluster.Namespace}
+}
+
+// OperatorClientCert returns the namespaced name for the client certificate
+// used by the Kubernetes operator
+func (r *PkiReconciler) OperatorClientCert() *types.NamespacedName {
+	if !r.pandaCluster.Spec.Configuration.TLS.RequireClientAuth {
+		return nil
+	}
+	return &types.NamespacedName{Name: r.pandaCluster.Name + "-" + OperatorClientCert, Namespace: r.pandaCluster.Namespace}
+}
+
+// Ensure will manage PKI for redpanda.vectorized.io custom resource
+func (r *PkiReconciler) Ensure(ctx context.Context) error {
+	if !r.pandaCluster.Spec.Configuration.TLS.KafkaAPIEnabled {
+		return nil
+	}
+
+	toApply := []resources.Resource{}
+
+	issuerRef := r.pandaCluster.Spec.Configuration.TLS.IssuerRef
+	// No cluster issuer is provided.
+	if issuerRef == nil {
+		selfSignedKey := r.issuerNamespacedName("selfsigned-issuer")
+		selfSignedIssuer := NewIssuer(r.Client,
+			r.scheme,
+			r.pandaCluster,
+			selfSignedKey,
+			cmapiv1.IssuerKind,
+			"",
+			r.logger)
+
+		rootCertificateKey := r.certNamespacedName("root-certificate")
+		rootCertificate := NewCertificate(r.Client,
+			r.scheme,
+			r.pandaCluster,
+			rootCertificateKey,
+			selfSignedIssuer.objRef(),
+			r.fqdn,
+			true,
+			r.logger)
+
+		// Kubernetes cluster issuer for Redpanda Operator - key provided in RedpandaCluster CR, else created
+		k8sClusterIssuerKey := r.issuerNamespacedName("root-issuer")
+		k8sClusterIssuer := NewIssuer(r.Client,
+			r.scheme,
+			r.pandaCluster,
+			k8sClusterIssuerKey,
+			cmapiv1.IssuerKind,
+			rootCertificate.Key().Name,
+			r.logger)
+
+		issuerRef = k8sClusterIssuer.objRef()
+		toApply = append(toApply, selfSignedIssuer, rootCertificate, k8sClusterIssuer)
+	}
+
+	// TODO: if a cluster issuer was provided, ensure that it comes with a CA (not self-signed). Perhaps create it otherwise.
+
+	// Redpanda cluster certificate for Kafka API - to be provided to each broker
+	certsKey := r.certNamespacedName(RedpandaNodeCert)
+	redpandaCert := NewCertificate(r.Client, r.scheme, r.pandaCluster, certsKey, issuerRef, r.fqdn, false, r.logger)
+
+	toApply = append(toApply, redpandaCert)
+
+	if r.pandaCluster.Spec.Configuration.TLS.RequireClientAuth {
+		// Certificate for external clients to call the Kafka API on any broker in this Redpanda cluster
+		certsKey = r.certNamespacedName(UserClientCert)
+		externalClientCert := NewCertificate(r.Client, r.scheme, r.pandaCluster, certsKey, issuerRef, r.fqdn, false, r.logger)
+
+		// Certificate for operator to call the Kafka API on any broker in this Redpanda cluster
+		certsKey = r.certNamespacedName(OperatorClientCert)
+		internalClientCert := NewCertificate(r.Client, r.scheme, r.pandaCluster, certsKey, issuerRef, r.fqdn, false, r.logger)
```

Comment:
this interface could be a simple builder pattern that makes it readable and extensible

```
cert = cert.NewBuilder()
                  .SetScheme(scheme)
                  .SetCertificateReference(issuerRef)
                  ...
                  .Build()
```

so that changing the arguments to something longer in the future would be a smaller change.

Reply:
thanks, opened an issue for this https://github.com/vectorizedio/redpanda/issues/817